File: TMB/Ex4.cpp
#include <TMB.hpp>

template <class Type> Type square(Type x){return x*x;}

template <class Type> vector<Type> square(vector<Type> x){return x*x;}

template<class Type>
Type objective_function<Type>::operator() ()
{
  DATA_INTEGER(Nyear)
  DATA_INTEGER(Nclass)
  DATA_VECTOR(Length)
  DATA_VECTOR(Weight)
  DATA_MATRIX(X);
  DATA_VECTOR(S)
  DATA_VECTOR(SurveyS)
  DATA_SCALAR(M)
  DATA_VECTOR(CWObs)
  DATA_MATRIX(CALObs);
  DATA_SCALAR(Neff)
  DATA_VECTOR(BioIndex)
  DATA_SCALAR(BioSig)
  DATA_INTEGER(Nproj)
  DATA_SCALAR(Fproj)

  // End of data section

  PARAMETER(dummy);
  PARAMETER(LogRbar);
  PARAMETER_VECTOR(LogNinit);
  PARAMETER_VECTOR(LogFullF);
  PARAMETER_VECTOR(Eps);

  matrix<Type> N(Nyear+Nproj+1,Nclass);
  matrix<Type> F(Nyear+Nproj,Nclass);
  matrix<Type> Z(Nyear+Nproj,Nclass);
  matrix<Type> CAL(Nyear+Nproj,Nclass);
  vector<Type> CW(Nyear+Nproj);
  vector<Type> BioPred(Nyear+Nproj);

  Type CALtot;

  Type Penal;
  Type LikeCatch;
  Type LikeBio;
  Type LikeCAL;
  Type obj_fun;

  // End of specifications section
  // =============================

  // First set F and Z by size-class (note that Fproj applies after year Nyear)
  for (int Iyear=0; Iyear<Nyear+Nproj; Iyear++)
   for (int Iclass=0;Iclass<Nclass;Iclass++)
    {
     if (Iyear < Nyear)
 	  F(Iyear,Iclass) = exp(LogFullF(Iyear))*S(Iclass);
 	 else
 	  F(Iyear,Iclass) = Fproj*S(Iclass);
	 Z(Iyear,Iclass) = M + F(Iyear,Iclass);

    }

  // Now set the N matrix
  for (int Iclass=0;Iclass<Nclass;Iclass++) N(0,Iclass) = exp(LogNinit(Iclass));
  for (int Iyear=0;Iyear<Nyear+Nproj;Iyear++)
   {
    // Catch-at-length
    CALtot = 0; CW(Iyear) = 0;
    for (int Iclass=0;Iclass<Nclass;Iclass++)
     {
      CAL(Iyear,Iclass) = F(Iyear,Iclass)/Z(Iyear,Iclass)*N(Iyear,Iclass)*(1.0-exp(-Z(Iyear,Iclass)));
      CALtot += CAL(Iyear,Iclass);
      CW(Iyear) += Weight(Iclass)*CAL(Iyear,Iclass);
     }
    for (int Iclass=0;Iclass<Nclass;Iclass++) CAL(Iyear,Iclass) /= CALtot;

    // Numbers-at-age
    for (int Iclass=0;Iclass<Nclass;Iclass++)
     {
      N(Iyear+1,Iclass) = 0;
      for (int Jclass=0;Jclass<Nclass;Jclass++)
       N(Iyear+1,Iclass) += N(Iyear,Jclass)*exp(-Z(Iyear,Jclass))*X(Jclass,Iclass);
     }

    // Recruitment (watch for the index for Eps - and N)
    N(Iyear+1,0) += exp(LogRbar)*exp(Eps[Iyear]);
   }

  // Catch Likelihood
  Type SS = 0;
  for (int Iyear=0; Iyear<Nyear; Iyear++)
   SS += square(log(CWObs(Iyear)) - log(CW(Iyear)));
  LikeCatch = SS /(2.0*0.05*0.05);

  // Biomass predictions
  for (int Iyear=0; Iyear<(Nyear+Nproj); Iyear++)
   {
    BioPred(Iyear) = 0;
    for (int Iclass=0;Iclass<Nclass;Iclass++) BioPred(Iyear) += N(Iyear,Iclass)*SurveyS(Iclass)*Weight(Iclass);
   }
  Type Top = 0; Type Bot = 0; Type q;
  for (int Iyear=0; Iyear<Nyear; Iyear++)
   { Top += log(BioIndex(Iyear)/BioPred(Iyear)); Bot += 1.0; }
  q = exp(Top/Bot);

  // Likelihood
  SS = 0;
  for (int Iyear=0; Iyear<Nyear; Iyear++)
   SS += square(log(BioIndex(Iyear))-log(q*BioPred(Iyear)));
  LikeBio = SS/(2*BioSig*BioSig);

  // CAL Likelihood
  LikeCAL = 0;
  for (int Iyear=0; Iyear<Nyear; Iyear++)
   for (int Iclass=0;Iclass<Nclass;Iclass++)
    if (CALObs(Iyear,Iclass) > 0)
     LikeCAL -= Neff*CALObs(Iyear,Iclass)*log(CAL(Iyear,Iclass)/CALObs(Iyear,Iclass));

  // Recruitment penalty (include years after Nyear)
  Penal = 0;
  for (int Iyear=0; Iyear<(Nyear+Nproj); Iyear++)
   Penal += Eps(Iyear)*Eps(Iyear);
  Penal = Penal / (2.0*0.6*0.6);

  obj_fun = dummy*dummy + LikeCatch+LikeBio+LikeCAL+Penal;

  // Stuff to report
  REPORT(F);
  REPORT(Z);
  REPORT(N);
  REPORT(LikeCatch);
  REPORT(LikeBio);
  REPORT(LikeCAL);
  REPORT(Penal);
  REPORT(CAL);
  REPORT(CW);
  REPORT(BioPred);
  REPORT(obj_fun);

  return(obj_fun);
}

File: deprecated/ex4.R
#devtools::install_github('colemonnahan/adnuts', build_vignettes=TRUE)

# =================================================================================================================

MCMC <- function(model,covar,xxx,mult,Nreps=22000,Nburns=2000,Nevery=10)
{
  library(mvtnorm)
  
  Npars <- length(xxx)
  Nout <- Nreps
  Outputs <- matrix(0,nrow=Nout,ncol=Npars)
  ObjFn <- rep(NA,length=Nout)
  
  # Counters
  ichk <- 0
  jchk <- 0
  
  # Initialize
  Theta <- xxx
  DeltaInit <- 0.1
  Delta <- DeltaInit * xxx
  print(Theta)
  gold <- exp(-1*model$fn(Theta))
  print("gold")
  print(log(gold))
  
  Iout <- 0
  for (Irep in 1:Nreps)
  {
    
    # do for each parameter
    asave <- Theta
    Theta <- rmvnorm(1, mean=asave, sigma=covar*mult)
    g1 <- exp(-1*model$fn(Theta))
    #print(log(g1))
    
    RND <-runif(1,0,1)
    if (g1/gold > RND)
    { gold <- g1; ichk <- ichk+1 }
    else
    { Theta <- asave; jchk <- jchk + 1 } 
    
    if (Irep <= Nburns) 
      if (Irep %% Nevery == 0)
      {
        if (ichk > jchk)
          mult <- mult * 1.05
        else 
          mult <- mult * 0.95
        ichk <- 0
        jchk <- 0
      }  
    
    # end of current replicate
    if (Irep %% Nevery == 0)
      if (Irep > Nburns)
      { 
        Iout <- Iout + 1 
        Outputs[Iout,] <- Theta 
        ObjFn[Iout] <- log(gold)
      }
    if (Irep %% 1000 == 0) print(Irep)  
    
  }
  
  Outs <- list()
  Outs$Outs <- Outputs
  Outs$Outs <- Outs$Outs[1:Iout,]
  Outs$ObjFn <- ObjFn[1:Iout]
  return(Outs)
}

###############################
#################################################

DoMCMCAll <- function(model,VarCo)
 {  
  # Now consider mcmc sampling
  post1 <- MCMC(model,VarCo,model$env$last.par.best,mult=1.0,Nreps=1000000,Nburns=100000,Nevery=1000)
  post1a <- post1$Outs
  #save(post1a,file="post1.RData")

  # Now consider mcmc sampling (NUTS)
  #library(adnuts)
  #mcmcout <- sample_tmb(obj=model,seeds=1:3,init=list(list(model$env$last.par.best)), iter=20000,chains=1,algorithm="NUTS")
  #post2 <- extract_samples(mcmcout)
  #save(post2,file="F:\\post2.RData")
  #launch_shinytmb(mcmcout)
}  

################################################################################

MCMCSumm <- function(file,best,Nyear,data,map,parameters)
{
  # Load the parameter file
  load(file=file)

  # Set up for graphs
  par(mfrow=c(2,2))

  Nsim <- length(post1a[,1])
  Npar <- length(post1a[1,])
 
  # Good to check (compare with the sdreport)
  for (II in 1:Npar)  
   cat(II,best[II],mean(post1a[,II]),sd(post1a[,II]),"\n") 

  # Extract biomass and plot
  Biomass <- matrix(0,nrow=Nsim,ncol=Nyear+1)
  for (Isim in 1:Nsim)
   { xx <- model$fn(post1a[Isim,]); Biomass[Isim,] <- model$report()$BioPred[1:(Nyear+1)]; }  
  quant <- matrix(0,nrow=5,ncol=Nyear+1)
  for (Iyear in 1:(Nyear+1))
   quant[,Iyear] <- quantile(Biomass[,Iyear],probs=c(0.05,0.25,0.5,0.75,0.95))  
  Years <- 1:(Nyear+1)
  
  # Plot of biomass posterior
  ymax <- max(quant)
  plot(Years,quant[3,],xlab="Year",ylab="Biomass",ylim=c(0,ymax),type='l')
  xx <- c(Years,rev(Years))
  yy <- c(quant[1,],rev(quant[5,]))
  polygon(xx,yy,col="gray10")
  xx <- c(Years,rev(Years))
  yy <- c(quant[2,],rev(quant[4,]))
  polygon(xx,yy,col="gray90")
  lines(Years,quant[3,],lwd=3,lty=3)
  
  # Projection    
  # ==========
  Biomass <- matrix(0,nrow=Nsim,ncol=Nyear+20)

  # profile of probability of recovery with time  
  Probs <- rep(0,101)
  fs <- rep(0,101)
  for (IP in 1:101)
   {  
    fs[IP] <- (IP-1)*0.015
    data$Fproj=fs[IP]
    model <- MakeADFun(data, parameters, DLL="Ex4",silent=T,map=map,hessian=T)
    for (Isim in 1:Nsim)
     { xx <- model$fn(post1a[Isim,]); Biomass[Isim,] <- model$report()$BioPred[1:(Nyear+20)]; }  
    Probs[IP] <- sum(Biomass[,Nyear+20]>1000)/Nsim
   }  
  plot(fs,Probs,xlab="Fishing mortality",ylab="Probability > 1000t",ylim=c(0,1),type='l',lwd=2,lty=1) 

  # Bisection to find desire recovery probability
  Fmin <- 0
  Fmax <- 1
  for (II in 1:20)
   { 
    Fbar <- (Fmin+Fmax)/2.0
    data$Fproj=Fbar
    model <- MakeADFun(data, parameters, DLL="Ex4",silent=T,map=map,hessian=T)
    for (Isim in 1:Nsim)
    { xx <- model$fn(post1a[Isim,]); Biomass[Isim,] <- model$report()$BioPred[1:(Nyear+20)]; }  
    Probs <- sum(Biomass[,Nyear+20]>1000)/Nsim
    if (Probs > 0.5) Fmin <- Fbar else Fmax <- Fbar;
   }  
  cat(Fbar,Probs,"\n")
  points(Fbar,Probs,pch=16)
  abline(h=0.5,lty=2)
  abline(v=Fbar,lty=2)
}  

################################################################################


require(here)
require(TMB)

################################################################################

Nyear <- scan("ex4.dat",skip=1,n=1,quiet=T)
Nclass <- scan("ex4.dat",skip=3,n=1,quiet=T)
Length <- scan("ex4.dat",skip=5,n=Nclass,quiet=T)
Weight <- scan("ex4.dat",skip=7,n=Nclass,quiet=T)
X <- matrix(scan("ex4.dat",skip=9,n=Nclass*Nclass,quiet=T),ncol=Nclass,byrow=T)
M <- scan("ex4.dat",skip=19,n=1,quiet=T)
CWObs <- scan("ex4.dat",skip=22,n=Nyear,quiet=T)
CALObs <- matrix(scan("ex4.dat",skip=49,n=Nyear*Nclass,quiet=T),ncol=Nclass,byrow=T)
Neff <- scan("ex4.dat",skip=75,n=1,quiet=T)
BioIndex <- scan("ex4.dat",skip=78,n=Nyear,quiet=T)
BioSig <- scan("ex4.dat",skip=104,n=1,quiet=T)

# Selectivity parameter (convert to selectivity)
S50 <- scan("ex4.dat",skip=15,n=2,quiet=T)[1]
S95 <- scan("ex4.dat",skip=15,n=2,quiet=T)[2]
SS50 <- scan("ex4.dat",skip=17,n=2,quiet=T)[1]
SS95 <- scan("ex4.dat",skip=17,n=2,quiet=T)[2]
S <- rep(0,Nclass); SurveyS <- rep(0,Nclass)
for(Iclass in 1:Nclass)
 {
  S[Iclass] <- 1.0/(1+exp(-log(19.0)*(Length[Iclass]-S50)/(S95-S50)))
  SurveyS[Iclass] <- 1.0/(1+exp(-log(19.0)*(Length[Iclass]-SS50)/(SS95-SS50)))
 }  

# Normalize the catcha-at-length data
for (Iyear in 1:Nyear)
 {
  Total <- 0
  for (Iclass in 1:Nclass) Total <- Total + CALObs[Iyear,Iclass]
  for (Iclass in 1:Nclass) CALObs[Iyear,Iclass] = CALObs[Iyear,Iclass]/Total
 }  

################################################################################
# Hint Nproj = 0 for basic estimation
# Basic estimation
################################################################################

# Set the parameters to initial values
LogRbar <- scan("ex4.pin",skip=1,n=1,quiet=T)
LogNinit <- scan("ex4.pin",skip=3,n=Nclass,quiet=T)
LogFullF <- scan("ex4.pin",skip=5,n=Nyear,quiet=T)
Eps <- scan("ex4.pin",skip=9,n=Nyear,quiet=T)
Fproj <- 0

# Data vector
data <- list(Nyear=Nyear,Nclass=Nclass,Length=Length,Weight=Weight,X=X,S=S,SurveyS=SurveyS,M=M,
             CWObs=CWObs,CALObs=CALObs,Neff=Neff,BioIndex=BioIndex,BioSig=BioSig,Nproj=0,Fproj=0)
parameters <- list(dummy=0,LogRbar=LogRbar,LogNinit=LogNinit,LogFullF=LogFullF,Eps=Eps)
# When I was testing the code
map<-list(LogRbar=factor(NA),LogNinit=rep(factor(NA),Nclass),LogFullF=rep(factor(NA),Nyear),
          Eps=rep(factor(NA),Nyear))
# Estimate everything
map<-list(dummy=factor(NA))

#print(data)
#print(parameters)

compile(here("TMB","Ex4.cpp"))
dyn.load(here("TMB","Ex4"))

model <- MakeADFun(data, parameters, DLL="Ex4",silent=T,map=map,hessian=T)

# test code - for checking for minimization
xx <- model$fn(model$env$last.par)
#print(model$par)
print(model$report())
#cat(model$report()$obj_fun,model$report()$LikeCatch,model$report()$LikeCAL,model$report()$LikeBio,model$report()$Penal,"\n")
fit <- nlminb(model$par, model$fn, model$gr, control=list(eval.max=100000,iter.max=1000))
best <- model$env$last.par.best
print(best)
rep <- sdreport(model)
print(rep)
cat(model$report()$obj_fun,model$report()$LikeCatch,model$report()$LikeCAL,model$report()$LikeBio,model$report()$Penal,"\n")

################################################################################
# Hint Nproj = 20 for projections
# Basic estimation
################################################################################

Nproj = 20
compile("Ex4.cpp")
dyn.load(dynlib("Ex4"))

# Set the parameters to initial values
LogRbar <- scan("ex4.pin",skip=1,n=1,quiet=T)
LogNinit <- scan("ex4.pin",skip=3,n=Nclass,quiet=T)
LogFullF <- scan("ex4.pin",skip=5,n=Nyear,quiet=T)
Eps <- scan("ex4.pin",skip=9,n=Nyear,quiet=T)
Eps <- c(Eps,rep(0,Nproj))

# Data vector
data <- list(Nyear=Nyear,Nclass=Nclass,Length=Length,Weight=Weight,X=X,S=S,SurveyS=SurveyS,M=M,
             CWObs=CWObs,CALObs=CALObs,Neff=Neff,BioIndex=BioIndex,BioSig=BioSig,Nproj=Nproj,Fproj=0)
parameters <- list(dummy=0,LogRbar=LogRbar,LogNinit=LogNinit,LogFullF=LogFullF,Eps=Eps)
# When I was testing the code
map<-list(LogRbar=factor(NA),LogNinit=rep(factor(NA),Nclass),LogFullF=rep(factor(NA),Nyear),
          Eps=rep(factor(NA),Nyear))
# Estimate everything
map<-list(dummy=factor(NA))

#print(data)
#print(parameters)

# Note hessian=T so we get the Hessian matrix
model <- MakeADFun(data, parameters, DLL="Ex4",silent=T,map=map,hessian=T)

# test code - for checking for minimization
xx <- model$fn(model$env$last.par)
#print(model$par)
#print(model$report())
#cat(model$report()$obj_fun,model$report()$LikeCatch,model$report()$LikeCAL,model$report()$LikeBio,model$report()$Penal,"\n")
fit <- nlminb(model$par, model$fn, model$gr, control=list(eval.max=100000,iter.max=1000))
best <- model$env$last.par.best
print(best)
print(model$fn(best))
cat(model$report()$obj_fun,model$report()$LikeCatch,model$report()$LikeCAL,model$report()$LikeBio,model$report()$Penal,"\n")
VarCo <- solve(model$he())
rep <- sdreport(model)
print(rep)
print(model$report()$obj_fun)
# Check for Hessian
print(sqrt(diag(VarCo)))
#AAA

# Now consider mcmc sampling and provide a posterio for By
# ========================================================
#DoMCMCAll(model,VarCo)

# Now do the projections
# ==========================
MCMCSumm("post1.RData",model$env$last.par.best,Nyear,data,map,parameters)

  